lib/upnp/igd_service.go

```diff
@@ -50,6 +50,39 @@ type IGDService struct {
 	URL       string
 	URN       string
 	LocalIP   net.IP
+	PinholeID uint16
+}
+
+// TryAddPinhole adds an IPv6 pinhole in accordance to http://upnp.org/specs/gw/UPnP-gw-WANIPv6FirewallControl-v1-Service.pdf
+// We just use the same external and internal port
+func (s *IGDService) TryAddPinhole(ctx context.Context, protocol nat.Protocol, port int, description string, duration time.Duration) (int, error) {
+	var protoNumber int
+	if protocol == nat.TCP {
+		protoNumber = 6
+	} else if protocol == nat.UDP {
+		protoNumber = 17
+	}
+
+	tpl := `<u:AddPinhole xmlns:u="%s">
+	<RemoteHost>::/0</RemoteHost>
+	<RemotePort>%d</RemotePort>
+	<Protocol>%d</Protocol>
+	<InternalPort>%d</InternalPort>
+	<InternalClient>%s</InternalClient>
+	<LeaseTime>%d</LeaseTime>
+	</u:AddPinhole>`
+
+	body := fmt.Sprintf(tpl, s.URN, port, protoNumber, port, s.LocalIP, duration/time.Second)
+
+	response, err := soapRequest(ctx, s.URL, s.URN, "AddPinhole", body)
+	if err != nil && duration > 0 {
+		envelope := &soapErrorResponse{}
+		if unmarshalErr := xml.Unmarshal(response, envelope); unmarshalErr != nil {
+			return port, unmarshalErr
+		}
```

Comment:
Why go through the trouble of unmarshalling when you discard the result and only pass out the error if it failed?

And I guess it is common to re-use the `err` variable name here, no need to separate it if you're only returning from inside the conditional.

Reply:
Hm, you're right, this was copied from AddPortMapping but there's currently no need to actually parse the error response in AddPinhole (since there is no case like below where we can just retry with different options). I have removed it now.